ARROW-16000: [C++][Python] Dataset: Alternative implementation for adding transcoding function option to CSV scanner

joosthooz opened the pull request:
This is an alternative version of https://github.com/apache/arrow/pull/13709, to compare what the best approach is.

Instead of extending the C++ ReadOptions struct with an `encoding` field, this implementations adds a python version of the ReadOptions object to both `CsvFileFormat` and `CsvFragmentScanOptions`. The reason it is needed in both places, is to prevent these kinds of inconsistencies:
```
>>> import pyarrow.dataset as ds
>>> import pyarrow.csv as csv
>>> ro =csv.ReadOptions(encoding='iso8859')
>>> fo = ds.CsvFileFormat(read_options=ro)
>>> fo.default_fragment_scan_options.read_options.encoding
'utf8'
```

lidavidm commented on cpp/src/arrow/python/io.h:
```diff
@@ -112,5 +112,9 @@ std::shared_ptr<::arrow::io::InputStream> MakeTransformInputStream(
 
+using StreamWrapFunc = std::function<Result<std::shared_ptr<io::InputStream>>(
+    std::shared_ptr<io::InputStream>)>;
+std::shared_ptr<StreamWrapFunc> MakeStreamTransformFunc(TransformInputStreamVTable vtable,
```
```suggestion
ARROW_PYTHON_EXPORT
std::shared_ptr<StreamWrapFunc> MakeStreamTransformFunc(TransformInputStreamVTable vtable,
```

"symbol not found" errors on Windows are usually because of a missing `_EXPORT` macro

joosthooz commented:
There is 1 failure in the CI, it seems to have been a strange time-out.

lidavidm reviewed (commented):
I suppose what's left is adding a test of the transcoder?

joosthooz commented:
I added a test by copying and modifying the one for the csv reader, but ran into the following problem:
The test for latin-1 encoding works fine. However, the UTF16 test fails because there seems to be something wrong with the schema detection. The equality test fails, even if the schemas seem to be identical from a python point of view (the diffs that pytest prints are identical). Adding some printfs in the C++ code shows that the fields of `dataset.schema` seem to be empty:
```
Schema Equals():
this fp: 'S{Fn', other fp: 'S{Fna{@N};Fnb{@O};L}'
this field: '', that field: 'a: string'Schema Equals(): field 0 not equal
```
But this looks to me like a different problem with detecting the schema of a UTF16 encoded file. Should I try to create a reproducible example and file a new JIRA? Or is this something we should address here?
In the meantime, I removed the test in [47a3462](https://github.com/apache/arrow/pull/13820/commits/47a3462b756cf92594470cedcd0f56eaf6248016)

lidavidm commented:
That sounds like dataset inspection is being done without the transcoder actually being set. I think we do need the test to work. I would expect latin-1 happens to work because it happens that the header row has identical encoding between UTF-8 and latin-1.

joosthooz commented:
I added a test with a non-utf8 character in the column name (latin-1). That works. Looks like something a bit more specific to utf16. I'll investigate further.

joosthooz commented:
After having a better look, here's what seems to be happening:
- The first part of the test checks if parsing the file as binary still works. But that doesn't work for utf16 because the column names are not utf8. So parsing the column names into the schema fails (silently!).
- The second part tries to read the file, without specifying an encoding. It expects an exception. However, apparently the dataset reader has no problems with the null values every other character; it will just interpret it as a strange utf8 string.

I've removed those 2 additional checks, and just check if the data is transcoded properly. The 2nd check is still present in the new `test_column_names_encoding` test (that only tests latin-1)

joosthooz commented:
Hi guys, I think the current state should be ok, the failures seem unrelated to me. Is there anything else left for me to do?

lidavidm reviewed (approved):
Sorry for the delay. This looks good to me. @pitrou what do you think about the approach here?

pitrou commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1171,6 +1178,10 @@ cdef class CsvFileFormat(FileFormat):
+        # The encoding field in ReadOptions does not exist in the C++ struct.
+        # We need to store it here and override it when reading
+        # default_fragment_scan_options.read_options
+        public ReadOptions read_options_py
```
This does not need to be visible to the user, how about renaming it to stress it's an internal detail?
```suggestion
        public ReadOptions _read_options_py
```

joosthooz replied:
Thanks! changed in [b9982c8](https://github.com/apache/arrow/pull/13820/commits/b9982c8d28d898bbf506135bf53bad629728afef)

pitrou commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1251,6 +1266,9 @@ cdef class CsvFragmentScanOptions(FragmentScanOptions):
         CCsvFragmentScanOptions* csv_options
+        # The encoding field in ReadOptions does not exist in the C++ struct.
+        # We need to store it here and override it when reading read_options
+        ReadOptions read_options_py
```
Same here.

joosthooz replied:
Changed in [b9982c8](https://github.com/apache/arrow/pull/13820/commits/b9982c8d28d898bbf506135bf53bad629728afef)

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -3130,6 +3130,58 @@ def test_csv_fragment_options(tempdir, dataset_reader):
+def test_encoding(tempdir, dataset_reader):
+    path = str(tempdir / 'test.csv')
+
+    for encoding, input_rows, expected_table in [
```
`expected_table` here is unused.

joosthooz replied:
Thanks for catching that, it's been removed in [b3ac697](https://github.com/apache/arrow/pull/13820/commits/b3ac6979b0e78c96401dfe03298003287d1f9b7e)

pitrou reviewed (commented):
The approach looks fine to me. Just a couple comments.

joosthooz commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1278,11 +1297,18 @@ cdef class CsvFragmentScanOptions(FragmentScanOptions):
     def read_options(self, ReadOptions read_options not None):
         self.csv_options.read_options = deref(read_options.options)
+        self.read_options_py = read_options
+        if read_options.encoding != 'utf8':
```
I've added a lookup to deal with this in [1e621fa](https://github.com/apache/arrow/pull/13820/commits/1e621fa0c12493798c98447043eafde2ca250dac)

joosthooz commented:
Anything I can do to help move this forward? The failure seems unrelated to me (`cat: r/check/arrow.Rcheck/00install.out: No such file or directory`)
